mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp

```diff
@@ -2922,6 +2922,178 @@ convertDeclareTargetAttr(Operation *op, mlir::omp::DeclareTargetAttr attribute,
+        return convertOmpSections(*op, builder, moduleTranslation);
+      })
+      .Case([&](omp::SingleOp op) {
+        return convertOmpSingle(op, builder, moduleTranslation);
+      })
+      .Case([&](omp::TeamsOp op) {
+        return convertOmpTeams(op, builder, moduleTranslation);
+      })
+      .Case([&](omp::TaskOp op) {
+        return convertOmpTaskOp(op, builder, moduleTranslation);
+      })
+      .Case([&](omp::TaskGroupOp op) {
+        return convertOmpTaskgroupOp(op, builder, moduleTranslation);
+      })
+      .Case<omp::YieldOp, omp::TerminatorOp, omp::ReductionDeclareOp,
+            omp::CriticalDeclareOp>([](auto op) {
+        // `yield` and `terminator` can be just omitted. The block structure
+        // was created in the region that handles their parent operation.
+        // `reduction.declare` will be used by reductions and is not
+        // converted directly, skip it.
+        // `critical.declare` is only used to declare names of critical
+        // sections which will be used by `critical` ops and hence can be
+        // ignored for lowering. The OpenMP IRBuilder will create unique
+        // name for critical section names.
+        return success();
+      })
+      .Case([&](omp::ThreadprivateOp) {
+        return convertOmpThreadprivate(*op, builder, moduleTranslation);
+      })
+      .Case<omp::DataOp, omp::EnterDataOp, omp::ExitDataOp, omp::UpdateDataOp>(
+          [&](auto op) {
+            return convertOmpTargetData(op, builder, moduleTranslation);
+          })
+      .Case([&](omp::TargetOp) {
+        return convertOmpTarget(*op, builder, moduleTranslation);
+      })
+      .Case<omp::MapInfoOp, omp::DataBoundsOp, omp::PrivateClauseOp>(
+          [&](auto op) {
+            // No-op, should be handled by relevant owning operations e.g.
+            // TargetOp, EnterDataOp, ExitDataOp, DataOp etc. and then
+            // discarded
+            return success();
+          })
+      .Default([&](Operation *inst) {
+        return inst->emitError("unsupported OpenMP operation: ")
+               << inst->getName();
+      });
+}
+
+static LogicalResult
+convertInternalTargetOp(Operation *op, llvm::IRBuilderBase &builder,
+                        LLVM::ModuleTranslation &moduleTranslation) {
+  return convertCommonOperation(op, builder, moduleTranslation);
+}
+
+static LogicalResult
+convertTopLevelTargetOp(Operation *op, llvm::IRBuilderBase &builder,
+                        LLVM::ModuleTranslation &moduleTranslation) {
+  return llvm::TypeSwitch<Operation *, LogicalResult>(op)
+      .Case<omp::DataOp, omp::EnterDataOp, omp::ExitDataOp, omp::UpdateDataOp>(
+          [&](auto op) {
+            return convertOmpTargetData(op, builder, moduleTranslation);
+          })
+      .Case([&](omp::TargetOp) {
+        return convertOmpTarget(*op, builder, moduleTranslation);
+      })
+      // Skip omp ops that are not legal top level ops for the target device
```

Comment:
If I understand this correctly, this target region would be ignored:
```mlir
omp.task {
  omp.target {
  }
}
```
In general, by not processing the body of host operations, we will miss target regions nested within them. I don't know if there's a clean way to support that while following this approach.

Of course, if it isn't legal to nest target regions inside of any other constructs, this is a non-issue, but I haven't noticed anything like that in the spec. Maybe @mjklemm can enlighten us on that.

Reply:
Thinking about it a bit more, maybe it's not too difficult to support these cases. For those host ops that might contain a target region nested (immediately or not), we could just call `op.walk<omp::TargetOp>(...)` and trigger lowering for these target ops. The only problem this could have (I'm not sure, though) is that some arguments of that `omp.target` operation might not have been lowered to LLVM IR by then, since we would have skipped parent regions.

Reply:
```
program main
   integer :: A(10), B(10), C(10)

   do I = 1, 10
      A(I) = 1
      B(I) = 2
   end do
   !$omp target data map(to: A, B) map(alloc: C)
   !$omp target map(from: C)
   do I = 1, 10
      C(I) = A(I) + B(I) ! assigns 3, A:1 + B:2
   end do
   !$omp end target
   !$omp target update from(C) ! updates C device -> host
   !$omp end target data

   print *, C ! should be all 3's
end program
```

This is another use-case we need to be keep in mind, `target data` is a host construct that can contain a region that can have more than one `target`'s within it as well as other host related directives (e.g. the `update` in the example).

This already has a little bit of a bug in it just now that @TIFitis has looked into and attempted to address in: https://github.com/llvm/llvm-project/pull/85218

Reply:
This case does seem to be addressed in the above however, via `DataOp`, but thought I'd bring it up just to bring light to that particular use-case. 

It's possibly worth stating that I think `Exit`/`Enter`/`Update` are also host directives from my understanding and they do not own a region like `Data` that can contain device code in this case and likely should not be performed in device passes (there's been at least one error with this in the past). I don't think they can be invoked inside of a `target` directive/region either as far as the specification is concerned, but that would need some verification as it isn't something I've dug into too deeply.

Reply:
> If I understand this correctly, this target region would be ignored:
> 
> ```mlir
> omp.task {
>   omp.target {
>   }
> }
> ```
> 
> In general, by not processing the body of host operations, we will miss target regions nested within them. I don't know if there's a clean way to support that while following this approach.
> 
> Of course, if it isn't legal to nest target regions inside of any other constructs, this is a non-issue, but I haven't noticed anything like that in the spec. Maybe @mjklemm can enlighten us on that.

This exact structure is what both of my PRs (https://github.com/llvm/llvm-project/pull/85130 & https://github.com/llvm/llvm-project/pull/83966) will generate. According to the spec, this shouldn't be illegal (@mjklemm can correct me if I am wrong) - Quoting the spec

> When a target construct is encountered, a new target task is generated. The target task region
encloses the target region. 

This _target task_ could be interpreted as an implicit task but in the case of the `depend` clause (and I think the `nowait` clause as well) it needs an explicit task created (AFAICT as implied from the following statement in the spec 5.2 Section 13.8)
> If a depend clause is present, it is associated with the target task

Reply:
I put up a fix for handling inner regions before reading these comments, so keep that in mind if you look at the latest commit. But, yes the omp.task { omp.target { }} case was ignored in the original code. I added a case where the inner region of an op is traversed and calls the top level dispatch function recursively if it is an omp op. Probably not all operations with inner regions can contain top-level ops, but that should be handled earlier in the compiler, so we shouldn't have to check for illegal combinations.

Reply:
> Thinking about it a bit more, maybe it's not too difficult to support these cases. For those host ops that might contain a target region nested (immediately or not), we could just call `op.walk<omp::TargetOp>(...)` and trigger lowering for these target ops. The only problem this could have (I'm not sure, though) is that some arguments of that `omp.target` operation might not have been lowered to LLVM IR by then, since we would have skipped parent regions.

I think we are okay since omp.target is isolated from above, so I believe there shouldn't be any values that need to be translated (codegen seems to work for the omp.task { omp.target {}} case. I'm not sure if it is possible for the top level target data ops to have this issue.

Reply:
> Probably not all operations with inner regions can contain top-level ops, but that should be handled earlier in the compiler, so we shouldn't have to check for illegal combinations.

I agree, this would be way too late to realize that OpenMP operations were nested in an unsupported way.

> I think we are okay since omp.target is isolated from above, so I believe there shouldn't be any values that need to be translated (codegen seems to work for the omp.task { omp.target {}} case.

I would think that the region inside of omp.target should be translatable for that reason. My concern was more about arguments to that operation. For example, there's some processing of map arguments as part of `convertOmpTarget` that is done for both host and device. Will that still work if we haven't translated them to LLVM IR as part of lowering its enclosing region?
```mlir
%c1 = llvm.mlir.constant(1 : i64) : i64
%x = llvm.alloca %c1 x i32 : (i64) -> !llvm.ptr
omp.task {
  %map = omp.map_info var_ptr(%x : !llvm.ptr, i32) map_clauses(tofrom) capture(ByRef) -> !llvm.ptr
  omp.target map_entries(%map -> %arg0 : !llvm.ptr) {
  ^bb0(%arg0: !llvm.ptr):
    ...
    omp.terminator
  }
  omp.terminator
}
```
If that code makes this approach crash, then we would have to rethink the approach. We would need to create some dummy values, or somehow detect missing LLVM IR values and create them on demand. I think the former was basically what the early outlining did with the outlined function's arguments, if I remember correctly.

Reply:
The map arguments are already skipped in the common case, since they are supposed to be lowered by the omp.target op lowering and that lowering should not go beyond the omp.map_info operations, because that would include host code in the target codegen.

Reply:
> Of course, if it isn't legal to nest target regions inside of any other constructs, this is a non-issue, but I haven't noticed anything like that in the spec. Maybe @mjklemm can enlighten us on that.

It's allowed to nest `target` in other OpenMP constructs like `task`, `parallel`, `for`, etc.